### PR TITLE
feat: gtm server data transform

### PR DESCRIPTION
**Pull request by amliuyong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Review by zxkane — COMMENTED** [review]
@tyyzqmf Let's add this plug-in as a built-in third-party transformer plug-in.

File: src/data-pipeline/spark-etl/gradle/config/checkstyle/sun_checks.xml [hunk 1]
```diff
@@ -134,7 +134,7 @@
 
         <!-- Checks for whitespace                               -->
         <!-- See https://checkstyle.org/config_whitespace.html -->
-        <module name="EmptyForIteratorPad"/>
+        <!-- <module name="EmptyForIteratorPad"/> -->
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/data-pipeline/spark-etl/gradle/config/checkstyle/sun_checks.xml:137]: it is conflicted with IDE auto format 

IDE auto format always adds one space to then end `;`

below is the IDE auto format:

```java
 for (Iterator<String> ait = attrValue.fieldNames(); ait.hasNext(); )
```
Follow the rule, we need to change it as below: 
```
 for(Iterator<String> ait = attrValue.fieldNames(); ait.hasNext();)

```

**Comment by amliuyong** [comment]
> @tyyzqmf Let's add this plug-in as a built-in third-party transformer plug-in.

The transformer class is:

software.aws.solution.clickstream.gtm.GTMServerDataTransformer

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+public class ServerDataConverter {
+
+    private static UDF1<String, Row[]> convertGTMServerData() {
+        return (String value) -> {
+            try {
+                return getGenericRows(value);
+            } catch (Exception e) {
+                log.warn("cannot convert data: " + value + ", error: " + e.getMessage());
+                if (ContextUtil.isDebugLocal()) {
+                    e.printStackTrace();
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java:96]: use `log.error`

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+
+@Slf4j
+public class ServerDataConverter {
+
+    private static UDF1<String, Row[]> convertGTMServerData() {
+        return (String value) -> {
+            try {
+                return getGenericRows(value);
+            } catch (Exception e) {
+                log.warn("cannot convert data: " + value + ", error: " + e.getMessage());
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java:94]: should use `error`?

File: src/data-pipeline/spark-etl/src/test/java/software/aws/solution/clickstream/ETLRunnerTest.java [hunk 4]
```diff
@@ -323,10 +324,10 @@ public void should_executeTransformers_with_TransformerV2_3() throws IOException
 
         String dataDir = ContextUtil.getWarehouseDir();
 
-        String tableName1 = dataDir + "/" + TransformerV2.TABLE_ETL_USER_TRAFFIC_SOURCE + INCREMENTAL_SUFFIX ;
-        String tableName2 = dataDir + "/" + TransformerV2.TABLE_ETL_USER_DEVICE_ID + INCREMENTAL_SUFFIX;
-        String tableName3 = dataDir + "/" + TransformerV2.TABLE_ETL_USER_PAGE_REFERER + INCREMENTAL_SUFFIX;
-        String tableName4 = dataDir + "/" + TransformerV2.TABLE_ETL_USER_CHANNEL+ INCREMENTAL_SUFFIX;
+        String tableName1 = dataDir + "/" + TABLE_ETL_USER_TRAFFIC_SOURCE + INCREMENTAL_SUFFIX + "_v1";
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/test/java/software/aws/solution/clickstream/ETLRunnerTest.java:327]: should you extract `_v1` as constant?

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json [hunk 5]
```diff
@@ -0,0 +1,241 @@
+{
+  "appId" : "casetify_clickstream_analytics_aws",
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json:2]: change it

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  "ip" : "34.236.7.28",
+  "method" : "POST",
+  "path" : "/collect",
+  "platform" : null,
+  "rid" : "4a31fde2533e11dd2b0e7800720f6f86",
+  "server_ingest_time" : 1695261256000,
+  "source_type" : "http_server",
+  "timestamp" : "2023-09-21T01:54:16.554048761Z",
+  "ua" : null,
+  "uri" : "/collect?appId=casetify_clickstream_analytics_aws",
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json:16]: change appid

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+  "dataOut" : {
+    "gtmId" : "G-CCM0Y0Y1RC",
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json:18]: use fake one

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    "event_id" : "0-1695260713-2-448657059.1695257753",
+    "user_id" : null,
+    "event_name" : "PDP Clicks",
+    "ip" : "70.123.241.83",
+    "clientId" : "iKnHPfmF/8zea3TUOsTqbFpmAmhkTiNVL6Awhe5b43k=.1695257753",
+    "ua" : "Mozilla/5.0 (iPhone; CPU iPhone OS 16_6_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1",
+    "language" : "en-us",
+    "screenWidth" : 428,
+    "screenHeight" : 926,
+    "pageReferrer" : "https://www.casetify.com/product/screen-protector",
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json:29]: use example.com

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/DatasetUtil.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    public static final String MOBILE = "mobile";
+    public static final String MODEL = "model";
+    public static final String PLATFORM_VERSION = "platform_version";
+    public static final String ITEM_ID = "item_id";
+    public static final String PRICE = "price";
+    public static final String CLIENT_ID = "clientId";
+    public static final String SCREEN_WIDTH = "screenWidth";
+    public static final String SCREEN_HEIGHT = "screenHeight";
+    public static final String CLIENT_PLATFORM_VERSION = "clientPlatformVersion";
+    public static final String CLIENT_PLATFORM = "clientPlatform";
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/DatasetUtil.java:123]: changed

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/DatasetUtil.java [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+        );
+        saveFullDatasetToPath(pathInfo.getFull(), fullAggItemsDatasetRt);
+
+        log.info("fullAggItemsDataset count " + fullAggItemsDataset.count());
+        return fullAggItemsDataset;
+    }
+
+
+    private static String getPathForTable(final String tableName) {
+        String tableRegex = String.format("^(%s)|((%s|%s|(etl_[^/]+))(%s|%s)_v\\d+)$",
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/DatasetUtil.java:241]: defined the whole regex as const `TABLE_REGEX`

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java (hunk 2 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java:96]: changed.

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java (hunk 3 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/gtm/ServerDataConverter.java:94]: change to log.error()

File: src/data-pipeline/spark-etl/src/test/java/software/aws/solution/clickstream/ETLRunnerTest.java (hunk 4 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/test/java/software/aws/solution/clickstream/ETLRunnerTest.java:327]: changed.

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json (hunk 6 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json]: chaged.

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json (hunk 7 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json:18]: changed.

File: src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json (hunk 8 above)
Comment (amliuyong) [reply src/data-pipeline/spark-etl/src/test/resources/gtm-server/expected/test_convert_array_data.json]: changed.

**Comment by tyyzqmf** [comment]
> @tyyzqmf Let's add this plug-in as a built-in third-party transformer plug-in.

done

File: frontend/src/pages/plugins/create/CreatePlugin.tsx [hunk 11]
```diff
@@ -77,9 +77,13 @@ function Content() {
   >([]);
 
   const [s3Client, setS3Client] = useState<any>();
+  const [curDescription, setCurDescription] = useState<string>('');
   const [curPlugin, setCurPlugin] = useState<IPlugin>({
     name: '',
-    description: '',
+    description: {
```
Comment (zxkane) [inline frontend/src/pages/plugins/create/CreatePlugin.tsx:83]:
```ts
description: { [key: string]: string}
```

for more language support

File: src/control-plane/backend/lambda/api/model/plugin.ts [hunk 12]
```diff
@@ -18,7 +18,10 @@ export interface IPlugin {
   readonly prefix: string;
 
   readonly name: string;
-  readonly description: string;
+  readonly description: {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/model/plugin.ts:21]:
```ts
description: { [key: string]: string}
```

for more language support

File: src/control-plane/backend/lambda/api/test/api/ddb-mock.ts [hunk 13]
```diff
@@ -257,7 +257,10 @@ function dictionaryMock(ddbMock: any, name?: string): any {
             type: 'PLUGIN#BUILT-IN-1',
             prefix: 'PLUGIN',
             name: 'Transformer',
-            description: 'Description of Transformer',
+            description: {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/test/api/ddb-mock.ts:260]: In future, we might read the dict.json for generating mock data.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: frontend/src/pages/plugins/create/CreatePlugin.tsx [hunk 14]
```diff
@@ -397,9 +401,13 @@ function Content() {
                       return false;
                     }
                     setCurPlugin((prev) => {
+                      setCurDescription(e.detail.value);
                       return {
                         ...prev,
-                        description: e.detail.value,
+                        description: {
+                          'en-US': e.detail.value,
+                          'zh-CN': e.detail.value,
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/src/pages/plugins/create/CreatePlugin.tsx:409]: Avoid customer switching languages and edit description multiple times.

File: src/control-plane/backend/lambda/api/test/api/ddb-mock.ts (hunk 13 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/test/api/ddb-mock.ts:260]: #465

**Pull request merged**